### DISCUSSION TITLE
Prepared Http Client `PendingRequest`

### DISCUSSION THREAD

**Discussion by chu121su12** [laravel/framework#39934] [Q&A]
tldr: Similar to query builder that can be passed around before actually `->get()`-ing the result,
how to wrap `PendingRequest` so it can be send later? Ex:
```php
$http = \Http::withOptions([...])->willPost($url, $body);
// later
$response = $http->actualSend(); // options, method, url, body, etc already defined
```

***

I have this *simplified* http handling logic:
```php
$path = 0;
try {
    $response = \Http::withOptions([...])->post($url, $body);
} catch (\GuzzleHttp\Exception\RequestException $exception) {
    $path = 1;
} catch (\GuzzleHttp\Exception\ConnectException $exception) {
    $path = 2;
} catch (\Illuminate\Http\Client\ConnectionException $exception) {
    $curlErrorCode = preg_replace('/.+(\d+):.*/', '$1', $exception->getMessage());
    switch ($curlErrorCode) {
        case 7: $path = 2; break; // Failed to connect to <host> port <port>
        case 52: $path = 3; break; // Empty reply from server
        // more cases
        default: $path = 3;
    }
} catch (\Throwable $exception) {
    $path = 3;
}

switch ($path) {
    case 1: throw handleExceptionForPath1($exception);
    case 2: throw handleExceptionForPath2($exception);
    case 3: throw handleExceptionForPath3($exception);
    default: return handle($response);
}
```

There are a few api wtih different try-catch handling, and each api has multiple endpoint with different path, options, method and body format. With `$http` that can be passed around, I thought this can be simplified to a function defined once
```php
function handleEndpointA(PendingRequest $http) {
    try {
        $response = $http->actualSend();
    } catch (...) {
        // logic
    }
    // more logic with throw
    return $response;
}
```
Or is there any other way to do similar thing? I have extended this behavior with my own class, but if there are actual use of this logic, I'd like this to be implemented in core.
